src/solvers/refinement/string_constraint_generator.h

```diff
@@ -27,6 +27,53 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
 #include <util/constexpr.def>
 #include <solvers/refinement/string_constraint.h>
 
+/// Generation of fresh symbols of a given type
+class symbol_generatort final
+{
+public:
+  symbol_exprt
+  operator()(const irep_idt &prefix, const typet &type = bool_typet());
+
+private:
+  unsigned symbol_count = 0;
+};
+
+/// Correspondance between arrays and pointers string representations
+class array_poolt final
+{
+public:
+  const std::map<exprt, array_string_exprt> &get_arrays_of_pointers() const
+  {
+    return arrays_of_pointers;
+  }
+
+  exprt get_length(const array_string_exprt &s) const;
+
+  void insert(const exprt &pointer_expr, array_string_exprt &array);
+
+  array_string_exprt find(const exprt &pointer, const exprt &length);
+
+  array_string_exprt find(const refined_string_exprt &str);
+
+  /// Argument are passed to primitive as a struct with a pointer and length
+  /// field, this convert this struct to an array.
+  array_string_exprt of_argument(const exprt &arg);
+
+private:
+  // associate arrays to char pointers
+  std::map<exprt, array_string_exprt> arrays_of_pointers;
```

Comment:
Is there a reason not to use a `std::unordered_map` here? And indeed in all places except for those that require lexicographic sorting or stable iterators?

Reply:
No particular reason, I can change that